osutil/vfs/propagation.go

```diff
@@ -0,0 +1,265 @@
+}
+
+// MakeRecursivelyPrivate is a recursive version of MakePrivate.
+func (v *VFS) MakeRecursivelyPrivate(mountPoint string) error {
+	return v.makePrivate(mountPoint, flagRecurse)
+}
+
+// MakeUnbindable makes the mount point unbindable.
+//
+// A bind mount operation that uses an unbindable mount as source immediately
+// fails. A recursive bind mount operation ignores any unbindable mounts without
+// failure.
+func (v *VFS) MakeUnbindable(mountPoint string) error {
+	return v.makeUnbindable(mountPoint, 0)
+}
+
+// MakeRecursivelyUnbindable is a recursive version of MakeUnbindable.
+func (v *VFS) MakeRecursivelyUnbindable(mountPoint string) error {
+	return v.makeUnbindable(mountPoint, flagRecurse)
+}
+
+// prpagationChange applies a propagation change function fn at a given mount point.
+//
+// The string [op] is used to construct an error value in case [mountPoint] is
+// not a mount point.  The [recurse] flag causes the function to be called
+// recursively on all the descendants of the mount denoted by [mountPoint]. The
+// descendants are all the mounts whose parentID is that can be traced back to
+// the selected starting mount.
+func (v *VFS) propagationChange(mountPoint, op string, recurse bool, fn func(m *mount)) error {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+
+	return v.atMountPoint(mountPoint, op, func(pd pathDominator) error {
+		fn(pd.mount)
+
+		if recurse {
+			ancestorID := pd.mount.mountID
+			// TODO: iterate over all the children of a given mount.
+			for _, m := range v.mounts[:] {
+				if m.hasAncestor(ancestorID) {
+					fn(m)
+				}
+			}
+		}
+
+		return nil
+	})
+}
+
+// atMountPoint calls [fn] if [mountPoint] is an existing mount point.
+//
+// If [mountPoint] is not a mount point then a [fs.PathError] is returned,
+// encapsulating the given path [mountPoint], and operation [op].
+func (v *VFS) atMountPoint(mountPoint, op string, fn func(pd pathDominator) error) error {
+	// Find the mount dominating the mount point.
+	pd := v.pathDominator(mountPoint)
+
+	// If the VFS suffix is not empty then the given path is _not_ a mount point.
+	if pd.suffix != "" && pd.suffix != "." {
+		return &fs.PathError{Op: op, Path: mountPoint, Err: errNotMounted}
+	}
+
+	return fn(pd)
+}
+
+// unlockedSharedGroupCount countns the members of the shared peer group [id].
+func (v *VFS) unlockedSharedGroupCount(id GroupID) (count int) {
+	for _, m := range v.mounts {
+		if m.shared == id {
+			count++
+		}
+	}
+
+	return count
+}
+
+// maybePropagateMount propagates the mount m among the members of the parent's
+// peer group and their slaves. The function does nothing if the parent mount
+// is not shared.
+func (v *VFS) propagateMount(m *mount, skipFn func(m *mount) bool) {
+	if m.parent == nil {
+		panic("m.parent == nil")
+	}
+
+	if m.parent.shared == 0 {
+		panic("m.parent.shared == 0")
+	}
+
+	// TODO: use range syntax once we upgrade to go 1.23
+	v.peersAndSlavesOf(m.parent.shared)(func(peer *mount) bool {
```

Comment:
This is incorrect. We need to implement the complex iterator that is used in Linux to traverse first peers (in a particular order) than depth-first slaves.